Use enum class in webaudio/AudioContext

iidmsa opened the pull request:
#### 93d411576938dfec7e6bb7aa7146936080cd78ce
<pre>
Use enum class in webaudio/AudioContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=291398">https://bugs.webkit.org/show_bug.cgi?id=291398</a>
<a href="https://rdar.apple.com/problem/149029423">rdar://problem/149029423</a>

Reviewed by Chris Dumez.

Using Safer CPP constructs to avoid type confusion, and make it easier to pass over WebKit IPC.

* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::constructCommon):
(WebCore::AudioContext::willPausePlayback):
(WebCore::AudioContext::willBeginPlayback):
(WebCore::AudioContext::mediaCanStart):
* Source/WebCore/Modules/webaudio/AudioContext.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setAudioContextRestrictions):

Canonical link: <a href="https://commits.webkit.org/293764@main">https://commits.webkit.org/293764@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cc0e6fcc35ae073f595253ef38ab8002c07df3da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99617 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19267 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9527 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104748 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50213 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101658 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19657 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27700 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75830 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32927 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102624 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14978 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89949 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56189 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14778 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7935 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49577 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84722 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8020 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107107 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26732 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19517 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84789 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27096 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86153 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84306 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21459 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28978 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6687 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/20523 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26672 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31875 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26491 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29803 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28057 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/webaudio/AudioContext.h:
```diff
@@ -164,7 +164,7 @@ class AudioContext final
     MediaUniqueIdentifier m_currentIdentifier;
 
-    BehaviorRestrictions m_restrictions { NoRestrictions };
+    BehaviorRestrictions m_restrictions { static_cast<BehaviorRestrictions>(BehaviorRestrictionFlags::NoRestrictions) };
```
This should be using an OptionSet.

iidmsa replied:
Fixed.

cdumez commented on Source/WebCore/Modules/webaudio/AudioContext.h:
```diff
@@ -89,15 +89,15 @@ class AudioContext final
+    using BehaviorRestrictions = OptionSet<BehaviorRestrictionFlags>;
+    OptionSet<BehaviorRestrictionFlags> behaviorRestrictions() const { return m_restrictions; }
+    void addBehaviorRestriction(BehaviorRestrictionFlags restriction) { m_restrictions.add(restriction); }
+    void removeBehaviorRestriction(BehaviorRestrictionFlags restriction) { m_restrictions.remove(restriction); }
```
Seems this should take in a `BehaviorRestrictions`

iidmsa replied:
Done.

cdumez commented on Source/WebCore/Modules/webaudio/AudioContext.h:
```diff
@@ -89,15 +89,15 @@ class AudioContext final
-    void removeBehaviorRestriction(BehaviorRestrictions restriction) { m_restrictions &= ~restriction; }
+    using BehaviorRestrictions = OptionSet<BehaviorRestrictionFlags>;
+    OptionSet<BehaviorRestrictionFlags> behaviorRestrictions() const { return m_restrictions; }
+    void addBehaviorRestriction(BehaviorRestrictionFlags restriction) { m_restrictions.add(restriction); }
```
Seems this should take in a `BehaviorRestrictions`

iidmsa replied:
Done.

cdumez commented on Source/WebCore/Modules/webaudio/AudioContext.h:
```diff
@@ -89,15 +89,15 @@ class AudioContext final
-    void addBehaviorRestriction(BehaviorRestrictions restriction) { m_restrictions |= restriction; }
-    void removeBehaviorRestriction(BehaviorRestrictions restriction) { m_restrictions &= ~restriction; }
+    using BehaviorRestrictions = OptionSet<BehaviorRestrictionFlags>;
+    OptionSet<BehaviorRestrictionFlags> behaviorRestrictions() const { return m_restrictions; }
```
We could return a `BehaviorRestrictions` here.

iidmsa replied:
Done.

cdumez commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -5169,20 +5169,22 @@ void Internals::resumeAllMediaPlayback()
 void Internals::setAudioContextRestrictions(AudioContext& context, StringView restrictionsString)
 {
-    auto restrictions = context.behaviorRestrictions();
-    context.removeBehaviorRestriction(restrictions);
```
We can keep the original code with the changes I suggested above.

cdumez commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -5169,20 +5169,22 @@ void Internals::resumeAllMediaPlayback()
-            restrictions |= AudioContext::RequirePageConsentForAudioStartRestriction;
+            newRestrictions.add(AudioContext::BehaviorRestrictionFlags::RequirePageConsentForAudioStartRestriction);
     }
-    context.addBehaviorRestriction(restrictions);
```
I think we can keep the original code with the changes I suggested.

cdumez commented on Source/WebCore/Modules/webaudio/AudioContext.h:
```diff
@@ -89,15 +89,15 @@ class AudioContext final
     // Restrictions to change default behaviors.
-    enum BehaviorRestrictionFlags {
+    enum class BehaviorRestrictionFlags : uint8_t {
         NoRestrictions = 0,
```
We should remove this value. It is no longer useful with an OptionSet.

cdumez commented on Source/WebCore/Modules/webaudio/AudioContext.h:
```diff
@@ -175,6 +175,17 @@ class AudioContext final
 } // WebCore
 
+namespace WTF {
+template<> struct EnumTraits<WebCore::AudioContext::BehaviorRestrictionFlags> {
```
@achristensen07 Do we still need EnumTraits?

cdumez commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -5169,20 +5169,22 @@ void Internals::resumeAllMediaPlayback()
+    AudioContext::BehaviorRestrictions newRestrictions;
 
     for (StringView restrictionString : restrictionsString.split(',')) {
         if (equalLettersIgnoringASCIICase(restrictionString, "norestrictions"_s))
```
We can remove this whole `if` case I think. It does nothing.

cdumez reviewed (commented):
Pretty close. Definitely an improvement over the original code.

cdumez reviewed (approved):
Very nice.

cdumez reviewed (changes requested):
Actually, one of the IPC test failures is related. I suspect you've need to defined the enum class in a serialization.in file maybe to fix that test (cc @achristensen07).

cdumez reviewed (approved):
Actually, this test is failing due to:
```
ALERT: FAIL - IPC Metadata is incorrect for types: WebCore::TextDrawingModeFlags
```

So it doesn't appear related to your change. It would be nice to re-upload your PR to give this bot a chance to re-run the test.

iidmsa committed:
Use enum class in webaudio/AudioContext
https://bugs.webkit.org/show_bug.cgi?id=291398
rdar://problem/149029423

Reviewed by Chris Dumez.

Using Safer CPP constructs to avoid type confusion, and make it easier to pass over WebKit IPC.

* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::constructCommon):
(WebCore::AudioContext::willPausePlayback):
(WebCore::AudioContext::willBeginPlayback):
(WebCore::AudioContext::mediaCanStart):
* Source/WebCore/Modules/webaudio/AudioContext.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setAudioContextRestrictions):

Canonical link: https://commits.webkit.org/293764@main
